src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs

```diff
@@ -0,0 +1,327 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+
+    internal class DictionarySerializationWriter : ISerializationWriter
+    {   
+        internal const string DefaultKey = "Key";
+        internal const string DefaultObjectKey = "Obj";
+
+        private readonly Stack<string> lastPrefix = new Stack<string>(); // Stores previous prefix for the property Key
+        private readonly Stack<long> lastIndex = new Stack<long>(); // Stores previous index for unnamed property
+        private string currentPrefix = string.Empty; // Prefix for the property name to distinguish names in depth
+        private long currentIndex = 1; // Index for DefaultKey or DefaultObj to provide a property name for properties or objects with no names
+
+        internal DictionarySerializationWriter()
+        {
+            this.AccumulatedDictionary = new Dictionary<string, string>();
+            this.AccumulatedMeasurements = new Dictionary<string, double>();
+        }
+
+        internal Dictionary<string, string> AccumulatedDictionary { get; }
+        
+        internal Dictionary<string, double> AccumulatedMeasurements { get; }
+
+        public void WriteProperty(string name, string value)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("This argument requires a non-empty value", nameof(name));
+            }
+
+            string key = this.GetKey(name);
+
+            if (!this.AccumulatedDictionary.ContainsKey(key))
+            {
+                this.AccumulatedDictionary[key] = value;
+            }
+        }
+
+        public void WriteProperty(string name, double? value)
```

Comment:
We may want to consider this as a metric, but there are several places and use-cases where it should be property but not metric... Storing it as Property for the time being.